utils/man/man1/rapl-info.1
.TH "RAPL\-INFO" "1" "2021-04-14" "powercap" "rapl\-info"
.SH "NAME"
.LP
rapl\-info \- get RAPL configurations
.SH "SYNPOSIS"
.LP
\fBrapl\-info\fP [\fIOPTION\fP]...
.SH "DESCRIPTION"
.LP
This utility is deprecated, use
.BR powercap\-info (1)
instead.
.LP
Prints out Intel Running Average Power Limit (RAPL) configurations.
.LP
Note that the \-p and \-z options are different from
.BR powercap\-info (1).
.LP
This software requires an Intel processor (Sandy Bridge or newer), Linux
kernel 3.13 or newer compiled with \fBCONFIG_POWERCAP\fR and
\fBCONFIG_INTEL_RAPL\fR enabled, and the \fBintel_rapl\fR kernel module to
be loaded.
.SH "OPTIONS"
.LP
.TP
\fB\-h,\fR \fB\-\-help\fR
Prints out the help screen
.TP
\fB\-v,\fR \fB\-\-verbose\fR
Print errors when files are not available
.TP
\fB\-p,\fR \fB\-\-zone\fR=\fIZONE\fP
The zone number (none by default; 0 by default if using \-z/\-\-subzone
and/or \-c/\-\-constraint).
Ending with a colon prevents output for subzones.
E.g., for zone 0, but not subzones:
.br
\fB\-p 0:\fP
.TP
\fB\-p,\fR \fB\-\-package\fR=\fIPACKAGE\fP
Deprecated, use \-\-zone instead
.TP
\fB\-z,\fR \fB\-\-subzone=\fR\fISUBZONE\fP
The subzone number (none by default)
.TP
\fB\-c,\fR \fB\-\-constraint\fR=\fICONSTRAINT\fP
The constraint number (none by default)
.TP
All remaining options below are mutually exclusive:
.TP
\fB\-n,\fR \fB\-\-nzones\fR
Print the number of zones found, or the number of subzones found if
\-p/\-\-zone is set
.LP
The following are zone-level arguments (\-z/\-\-subzone is optional):
.TP
\fB\-j,\fR \fB\-\-z\-energy\fR
Print zone energy counter
.TP
\fB\-J,\fR \fB\-\-z\-max\-energy\-range\fR
Print zone maximum energy counter range
.TP
\fB\-w,\fR \fB\-\-z\-power\fR
Print zone current power
.TP
\fB\-W,\fR \fB\-\-z\-max\-power\-range\fR
Print zone maximum current power range
.TP
\fB\-e,\fR \fB\-\-z\-enabled\fR
Print zone enable/disabled status
.TP
\fB\-x,\fR \fB\-\-z\-name\fR
Print zone name
.LP
The following are constraint-level arguments and require
\-c/\-\-constraint (\-z/\-\-subzone is optional):
.TP
\fB\-l,\fR \fB\-\-c\-power\-limit\fR
Print constraint power limit
.TP
\fB\-s,\fR \fB\-\-c\-time\-window\fR
Print constraint time window
.TP
\fB\-U,\fR \fB\-\-c\-max\-power\fR
Print constraint maximum allowed power
.TP
\fB\-u,\fR \fB\-\-c\-min\-power\fR
Print constraint minimum allowed power
.TP
\fB\-T,\fR \fB\-\-c\-max\-time\-window\fR
Print constraint maximum allowed time window
.TP
\fB\-t,\fR \fB\-\-c\-min\-time\-window\fR
Print constraint minimum allowed time window
.TP
\fB\-y,\fR \fB\-\-c\-name\fR
Print constraint name
.SH "EXAMPLES"
.LP
Note that \-p/\-\-zone=0 is used by default when also specifying
\-z/\-\-subzone and/or \-c/\-\-constraint, allowing for simpler commands
on single-socket systems.
.TP
\fBrapl\-info\fP
Print all RAPL zones.
.TP
\fBrapl\-info \-p 0\fP
Print only zone 0, which is usually named \fBpackage\-0\fR.
.TP
\fBrapl\-info \-p 0 \-z 1\fP
Print zone 0, subzone 1, which is usually the \fBuncore\fR or
\fBdram\fR subzone of \fBpackage\-0\fR, depending on the system.
.TP
\fBrapl\-info \-p 0 \-c 1\fP
Print zone 0, constraint 1, which is usually the \fBshort_term\fR
constraint for \fBpackage\-0\fR.
.TP
\fBrapl\-info \-p 0 \-j\fP
Print the energy counter for zone 0, which is usually named
\fBpackage\-0\fR.
.TP
\fBrapl\-info \-p 1 \-z 0 \-c 0 \-l\fP
Print the power limit for zone 1, subzone 0, constraint 0, which is
usually the \fBlong_term\fR constraint for the \fBcore\fR subzone of
\fBpackage\-1\fR (a multi-socket system).
.SH "REMARKS"
.LP
Some fields are optional and will only be printed if they are available
unless \-v/\-\-verbose is set.
.br
If no zone/constraint-specific outputs are requested, all available
zones and constraints will be shown.
.LP
Energy units: microjoules (uJ)
.br
Power units: microwatts (uW)
.br
Time units: microseconds (us)
.SH "BUGS"
.LP
The following are behavioral quirks due to the kernel interface or
abnormalities in some hardware.
They are not bugs in \fBrapl\-info\fR and should not be reported as such.
.LP
Values returned by the kernel sometimes lose accuracy from the actual
values stored in Model-Specific Registers due to integer rounding.
For example, the kernel may return a time window value of 7812 us, when in
fact the actual stored value is 7812.5 us.
Writing back a value returned by the kernel, e.g., using rapl\-set(1), may
cause unexpected changes to its value due to these rounding problems.
.LP
At the time of this writing, some features in the powercap interface are not
supported by RAPL.
The following options may result in ENOENT (No such file or directory) errors:
  \fB\-w/\-\-z\-power\fR
  \fB\-W/\-\-z\-max\-power\-range\fR
  \fB\-u/\-\-c\-min\-power\fR
  \fB\-T/\-\-c\-max\-time\-window\fR
  \fB\-t/\-\-c\-min\-time\-window\fR
.LP
On some systems, the constraint max_power_uw file, e.g.,
constraint_0_max_power_uw, is known to receive an ENODATA (No data
available) error from the kernel for subzones.
So even though the file exists, a value may not be printed for it.
.LP
It is possible for packages on multi-socket systems to be indexed out of
order by the kernel.
For example, the package at index 0 (\fB\-p 0\fR) could actually be named
\fBpackage\-1\fR while the package at index 1 (\fB\-p 1\fR) is named
\fBpackage\-0\fR.
.LP
Report bugs upstream at <https://github.com/powercap/powercap>
.SH "FILES"
.nf
\fI/sys/devices/virtual/powercap/intel\-rapl/*\fP
.nf
\fI/sys/class/powercap/intel\-rapl/*\fP
.fi
.SH "AUTHORS"
.nf
Connor Imes <connor.k.imes@gmail.com>
.fi
.SH "SEE ALSO"
.BR powercap\-info (1),
.BR powercap\-set (1),
.BR rapl\-set (1)

RELEASES.md
# Release Notes

## [v0.4.0] - 2021-04-25

### Fixed

* powercap.h: header documentation for 'powercap_*_file_get_name' functions' return behavior
* rapl-{info,set}: deprecation messages that specified non-existent alternatives

### Added

* Functions to support opening files in powercap.h:
  * 'powercap_control_type_file_open'
  * 'powercap_zone_file_open'
  * 'powercap_constraint_file_open'
* CMake helper to find powercap without relying on pkg-config

### Changed

* Binaries powercap-{info,set} now accept the control type argument without requiring a preceding -p flag
* Increased minimum CMake version from 2.8.12 to 3.1

### Deprecated

* Flag '-p/--control-type' in powercap-{info,set} - specify control type name as the first positional argument instead
* Add compiler deprecation attributes to deprecated functions


## [v0.3.1] - 2020-11-08

### Fixed

* [#6] powercap-common-test:test_snprintf_base_path fails on ppc64el architecture


## [v0.3.0] - 2020-08-15

### Added

* Support for top-level control type
  * Struct 'powercap_control_type', enum 'powercap_control_type_file', and associated functions in powercap.h:
    * 'powercap_control_type_get_enabled'
    * 'powercap_control_type_set_enabled'
    * 'powercap_control_type_file_get_name'
  * Functions in powercap-sysfs.h:
    * 'powercap_sysfs_control_type_get_enabled'
    * 'powercap_sysfs_control_type_set_enabled'
  * Functions in powercap-rapl.h:
    * 'powercap_rapl_control_is_supported'
    * 'powercap_rapl_control_is_enabled'
    * 'powercap_rapl_control_set_enabled'
  * Argument '--enabled' to powercap-{info,set} for getting/setting control type enabled field
* Function 'powercap_rapl_get_num_instances' in powercap-rapl.h (supersedes 'powercap_rapl_get_num_packages')
* Argument '--nconstraints' to powercap-info for getting the number of zone constraints

### Changed

* Increased minimum CMake version from 2.8.5 to 2.8.12 to support target_compile_definitions
* Updated 'powercap-{info,set}' man pages
* Argument '--zone' for powercap-set no longer required due to introduction of '--enabled' argument
* Disabled logging by default (in general, libraries shouldn't print output unless requested)

### Deprecated

* Binaries rapl-{info,set} - use powercap-{info,set} instead
* Interface powercap-rapl-sysfs.h - use powercap-sysfs.h directly instead
* Function 'powercap_rapl_get_num_packages' in powercap-rapl.h - use 'powercap_rapl_get_num_instances' instead

### Fixed

* Fixed [#4]: powercap-rapl now checks if parent zone is PACKAGE or PSYS


## [v0.2.0] - 2019-12-03

### Added

* This RELEASES.md file
* Multiarch support (use GNU standard installation directories)
* Additional documentation in README
* Function 'rapl_sysfs_zone_exists' in powercap-rapl-sysfs.h
* Command line long option '--zone' to rapl-info and rapl-set
* Functions to powercap-rapl-sysfs.h allowed by the powercap API that RAPL doesn't support (but could in the future)
* Options to rapl-info and rapl-set to support optional powercap features not currently implemented in RAPL

### Changed

* Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
* More pedantic man page source formatting
* RAPL sysfs 'pkg' parameter names changed to 'zone' to improve generality
* RAPL 'package' parameter names changed to 'id' to improve generality
* On failure, binaries now exit with positive error codes instead of negative values

### Deprecated

* Function 'rapl_sysfs_pkg_exists' in powercap-rapl-sysfs.h - use 'rapl_sysfs_zone_exists' instead
* Function 'rapl_sysfs_sz_exists' in powercap-rapl-sysfs.h - use 'rapl_sysfs_zone_exists' instead
* Command line long option '--package' for rapl-info and rapl-set - use '--zone' instead

### Removed

* Removed private symbol exports in shared object library (already patched in Debian)
* Undocumented (and unintended) --package long option from powercap-set

### Fixed

* Fixed [#1]: Kernel uses hexadecimal numbers in directory paths, not decimal


## [v0.1.1] - 2017-09-21

### Added

* Added stateless interfaces to sysfs
* Added powercap-info, powercap-set, rapl-info, and rapl-set binaries and man pages
* Added VERSION and SOVERSION to shared object libraries

### Changed

* Update license to use author as copyright holder


## v0.1.0 - 2017-06-09

### Added

* Initial public release

[v0.4.0]: https://github.com/powercap/powercap/compare/v0.3.1...v0.4.0
[v0.3.1]: https://github.com/powercap/powercap/compare/v0.3.0...v0.3.1
[v0.3.0]: https://github.com/powercap/powercap/compare/v0.2.0...v0.3.0
[v0.2.0]: https://github.com/powercap/powercap/compare/v0.1.1...v0.2.0
[v0.1.1]: https://github.com/powercap/powercap/compare/v0.1.0...v0.1.1
[#6]: https://github.com/powercap/powercap/issues/6
[#4]: https://github.com/powercap/powercap/issues/4
[#1]: https://github.com/powercap/powercap/issues/1


utils/rapl-info.c
/*
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Get RAPL values.
 *
 * @author Connor Imes
 * @date 2017-08-24
 */
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include "powercap-rapl-sysfs.h"
#include "util-common.h"

static void print_headers(uint32_t zone, uint32_t do_zone, uint32_t sz, int is_sz) {
  if (do_zone) {
    printf("Zone %"PRIu32"\n", zone);
  }
  if (is_sz) {
    indent(1);
    printf("Subzone %"PRIu32"\n", sz);
  }
}

static void analyze_constraint(uint32_t zone, uint32_t sz, int is_sz, uint32_t constraint, int verbose) {
  char name[MAX_NAME_SIZE];
  uint64_t val64;
  ssize_t sret;
  int ret;

  indent(is_sz + 1);
  printf("Constraint %"PRIu32"\n", constraint);

  sret = rapl_sysfs_constraint_get_name(zone, sz, is_sz, constraint, name, sizeof(name));
  ret = sret > 0 ? 0 : (int) sret;
  str_or_verbose(verbose, is_sz + 2, "name", name, ret);

  ret = rapl_sysfs_constraint_get_power_limit_uw(zone, sz, is_sz, constraint, &val64);
  u64_or_verbose(verbose, is_sz + 2, "power_limit_uw", val64, ret);

  ret = rapl_sysfs_constraint_get_time_window_us(zone, sz, is_sz, constraint, &val64);
  u64_or_verbose(verbose, is_sz + 2, "time_window_us", val64, ret);

  ret = rapl_sysfs_constraint_get_min_power_uw(zone, sz, is_sz, constraint, &val64);
  u64_or_verbose(verbose, is_sz + 2, "min_power_uw", val64, ret);

  ret = rapl_sysfs_constraint_get_max_power_uw(zone, sz, is_sz, constraint, &val64);
  u64_or_verbose(verbose, is_sz + 2, "max_power_uw", val64, ret);

  ret = rapl_sysfs_constraint_get_min_time_window_us(zone, sz, is_sz, constraint, &val64);
  u64_or_verbose(verbose, is_sz + 2, "min_time_window_us", val64, ret);

  ret = rapl_sysfs_constraint_get_max_time_window_us(zone, sz, is_sz, constraint, &val64);
  u64_or_verbose(verbose, is_sz + 2, "max_time_window_us", val64, ret);
}

static void analyze_zone(uint32_t zone, uint32_t sz, int is_sz, int verbose) {
  char name[MAX_NAME_SIZE];
  uint64_t val64;
  uint32_t val32;
  ssize_t sret;
  int ret;

  print_headers(0, 0, sz, is_sz);

  sret = rapl_sysfs_zone_get_name(zone, sz, is_sz, name, sizeof(name));
  ret = sret > 0 ? 0 : (int) sret;
  str_or_verbose(verbose, is_sz + 1, "name", name, ret);

  ret = rapl_sysfs_zone_get_enabled(zone, sz, is_sz, &val32);
  u64_or_verbose(verbose, is_sz + 1, "enabled", (uint64_t) val32, ret);

  ret = rapl_sysfs_zone_get_max_energy_range_uj(zone, sz, is_sz, &val64);
  u64_or_verbose(verbose, is_sz + 1, "max_energy_range_uj", val64, ret);

  ret = rapl_sysfs_zone_get_energy_uj(zone, sz, is_sz, &val64);
  u64_or_verbose(verbose, is_sz + 1, "energy_uj", val64, ret);

  ret = rapl_sysfs_zone_get_max_power_range_uw(zone, sz, is_sz, &val64);
  u64_or_verbose(verbose, is_sz + 1, "max_power_range_uw", val64, ret);

  ret = rapl_sysfs_zone_get_power_uw(zone, sz, is_sz, &val64);
  u64_or_verbose(verbose, is_sz + 1, "power_uw", val64, ret);

  for (val32 = 0; !rapl_sysfs_constraint_exists(zone, sz, is_sz, val32); val32++) {
    analyze_constraint(zone, sz, is_sz, val32, verbose);
  }
}

static void analyze_zone_recurse(uint32_t zone, int verbose) {
  uint32_t sz;
  print_headers(zone, 1, 0, 0);
  analyze_zone(zone, 0, 0, verbose);
  for (sz = 0; !rapl_sysfs_zone_exists(zone, sz, 1); sz++) {
    analyze_zone(zone, sz, 1, verbose);
  }
}

static void analyze_all_zones_recurse(int verbose) {
  uint32_t zone;
  for (zone = 0; !rapl_sysfs_zone_exists(zone, 0, 0); zone++) {
    analyze_zone_recurse(zone, verbose);
  }
}

static void print_num_zones(void) {
  uint32_t n = 0;
  while (!rapl_sysfs_zone_exists(n, 0, 0)) {
    n++;
  }
  printf("%"PRIu32"\n", n);
}

static void print_num_subzones(uint32_t zone) {
  uint32_t n = 0;
  while (!rapl_sysfs_zone_exists(zone, n, 1)) {
    n++;
  }
  printf("%"PRIu32"\n", n);
}

static const char short_options[] = "hvp:z:c:njJwWexlsUuTty";
static const struct option long_options[] = {
  {"help",                no_argument,        NULL, 'h'},
  {"verbose",             no_argument,        NULL, 'v'},
  {"package",             required_argument,  NULL, 'p'},
  {"zone",                required_argument,  NULL, 'p'},
  {"subzone",             required_argument,  NULL, 'z'},
  {"constraint",          required_argument,  NULL, 'c'},
  {"nzones",              no_argument,        NULL, 'n'},
  {"z-energy",            no_argument,        NULL, 'j'},
  {"z-max-energy-range",  no_argument,        NULL, 'J'},
  {"z-power",             no_argument,        NULL, 'w'},
  {"z-max-power-range",   no_argument,        NULL, 'W'},
  {"z-enabled",           no_argument,        NULL, 'e'},
  {"z-name",              no_argument,        NULL, 'x'},
  {"c-power-limit",       no_argument,        NULL, 'l'},
  {"c-time-window",       no_argument,        NULL, 's'},
  {"c-max-power",         no_argument,        NULL, 'U'},
  {"c-min-power",         no_argument,        NULL, 'u'},
  {"c-max-time-window",   no_argument,        NULL, 'T'},
  {"c-min-time-window",   no_argument,        NULL, 't'},
  {"c-name",              no_argument,        NULL, 'y'},
  {0, 0, 0, 0}
};

static void print_usage(void) {
  printf("\nThis utility is deprecated, use powercap-info instead.\n\n");
  printf("Usage: rapl-info [OPTION]...\n");
  printf("Options:\n");
  printf("  -h, --help                   Print this message and exit\n");
  printf("  -v, --verbose                Print errors when files are not available\n");
  printf("  -p, --zone=ZONE              The zone number (none by default; 0 by default if\n");
  printf("                               using -z/--subzone and/or -c/--constraint)\n");
  printf("                               Ending with a colon prevents output for subzones\n");
  printf("                               E.g., for zone 0, but not subzones: \"-p 0:\"\n");
  printf("      --package=PACKAGE        Deprecated, use --zone instead\n");
  printf("  -z, --subzone=SUBZONE        The subzone number (none by default)\n");
  printf("  -c, --constraint=CONSTRAINT  The constraint number (none by default)\n");
  printf("All remaining options below are mutually exclusive:\n");
  printf("  -n, --nzones                 Print the number of zones found, or the number of\n");
  printf("                               subzones found if -p/--zone is set\n");
  printf("The following are zone-level arguments (-z/--subzone is optional):\n");
  printf("  -j, --z-energy               Print zone energy counter\n");
  printf("  -J, --z-max-energy-range     Print zone maximum energy counter range\n");
  printf("  -w, --z-power                Print zone current power\n");
  printf("  -W, --z-max-power-range      Print zone maximum current power range\n");
  printf("  -e, --z-enabled              Print zone enabled/disabled status\n");
  printf("  -x, --z-name                 Print zone name\n");
  printf("The following are constraint-level arguments and require -c/--constraint (-z/--subzone is optional):\n");
  printf("  -l, --c-power-limit          Print constraint power limit\n");
  printf("  -s, --c-time-window          Print constraint time window\n");
  printf("  -U, --c-max-power            Print constraint maximum allowed power\n");
  printf("  -u, --c-min-power            Print constraint minimum allowed power\n");
  printf("  -T, --c-max-time-window      Print constraint maximum allowed time window\n");
  printf("  -t, --c-min-time-window      Print constraint minimum allowed time window\n");
  printf("  -y, --c-name                 Print constraint name\n");
  printf("\nSome fields are optional and will only be printed if they are available unless -v/--verbose is set.\n");
  printf("If no zone/constraint-specific outputs are requested, all available zones and constraints will be shown.\n");
  printf("\nEnergy units: microjoules (uJ)\n");
  printf("Power units: microwatts (uW)\n");
  printf("Time units: microseconds (us)\n");
}

static void print_common_help(void) {
  printf("Considerations for common errors:\n");
  printf("- Ensure that the intel_rapl kernel module is loaded\n");
  printf("- Some files may simply not exist\n");
  printf("- On some systems, the kernel always returns an error when reading constraint max power (-U/--c-max-power)\n");
}

int main(int argc, char** argv) {
  u32_param zone = {0, 0};
  u32_param subzone = {0, 0};
  u32_param constraint = {0, 0};
  int recurse = 1;
  int verbose = 0;
  int unique_set = 0;
  int c;
  int cont = 1;
  int ret = 0;
  uint64_t val64;
  uint32_t val32;
  char name[MAX_NAME_SIZE];

  /* Parse command-line arguments */
  while (cont) {
    c = getopt_long(argc, argv, short_options, long_options, NULL);
    switch (c) {
    case -1:
      cont = 0;
      break;
    case 'h':
      print_usage();
      return EXIT_SUCCESS;
    case 'v':
      verbose = 1;
      break;
    case 'p':
      recurse = get_recurse(optarg);
      ret = set_u32_param(&zone, optarg, &cont);
      break;
    case 'z':
      ret = set_u32_param(&subzone, optarg, &cont);
      break;
    case 'c':
      ret = set_u32_param(&constraint, optarg, &cont);
      break;
    case 'n':
    case 'j':
    case 'J':
    case 'w':
    case 'W':
    case 'e':
    case 'x':
    case 'l':
    case 's':
    case 'U':
    case 'u':
    case 'T':
    case 't':
    case 'y':
      if (unique_set) {
        fprintf(stderr, "Only one of -n/--nzones, a zone-level argument, or a constraint-level argument is allowed at a time\n");
        cont = 0;
        ret = -EINVAL;
        break;
      }
      unique_set = c;
      break;
    case '?':
    default:
      cont = 0;
      ret = -EINVAL;
      break;
    }
  }

  /* Verify argument combinations */
  if (ret) {
    fprintf(stderr, "Unknown or duplicate arguments\n");
  } else {
    switch (unique_set) {
    case 'n':
      if (subzone.set || constraint.set) {
        fprintf(stderr, "-n/--nzones cannot be used with -z/--subzone or -c/--constraint\n");
        ret = -EINVAL;
      }
      break;
    case 'j':
    case 'J':
    case 'w':
    case 'W':
    case 'e':
    case 'x':
      if (constraint.set) {
        fprintf(stderr, "-c/--constraint cannot be set for zone-level arguments\n");
        ret = -EINVAL;
      }
      break;
    case 'l':
    case 's':
    case 'U':
    case 'u':
    case 'T':
    case 't':
    case 'y':
      if (!constraint.set) {
        fprintf(stderr, "-c/--constraint must be set for constraint-level arguments\n");
        ret = -EINVAL;
      }
      break;
    }
  }
  if (ret) {
    print_usage();
    return EXIT_FAILURE;
  }

  /* Check if zone/subzone/constraint exist */
  if (rapl_sysfs_zone_exists(zone.val, 0, 0)) {
    fprintf(stderr, "Zone does not exist\n");
    ret = -EINVAL;
  } else if (subzone.set && rapl_sysfs_zone_exists(zone.val, subzone.val, 1)) {
    fprintf(stderr, "Subzone does not exist\n");
    ret = -EINVAL;
  } else if (constraint.set && rapl_sysfs_constraint_exists(zone.val, subzone.val, subzone.set, constraint.val)) {
    fprintf(stderr, "Constraint does not exist\n");
    ret = -EINVAL;
  }
  if (ret) {
    print_common_help();
    return EXIT_FAILURE;
  }

  /* Perform requested action */
  if (unique_set) {
    switch (unique_set) {
    case 'n':
      /* Print number of zones or subzones */
      if (zone.set) {
        print_num_subzones(zone.val);
      } else {
        print_num_zones();
      }
      break;
    case 'j':
      /* Get zone energy */
      if (!(ret = rapl_sysfs_zone_get_energy_uj(zone.val, subzone.val, subzone.set, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get zone energy");
      }
      break;
    case 'J':
      /* Get zone max energy range */
      if (!(ret = rapl_sysfs_zone_get_max_energy_range_uj(zone.val, subzone.val, subzone.set, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get zone max energy range");
      }
      break;
    case 'w':
      /* Get zone power */
      if (!(ret = rapl_sysfs_zone_get_power_uw(zone.val, subzone.val, subzone.set, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get zone power");
      }
      break;
    case 'W':
      /* Get zone max power range */
      if (!(ret = rapl_sysfs_zone_get_max_power_range_uw(zone.val, subzone.val, subzone.set, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get zone max power range");
      }
      break;
    case 'e':
      /* Get zone enabled */
      if (!(ret = rapl_sysfs_zone_get_enabled(zone.val, subzone.val, subzone.set, &val32))) {
        printf("%"PRIu32"\n", val32);
      } else {
        perror("Failed to get zone enabled");
      }
      break;
    case 'x':
      /* Get zone name */
      if (rapl_sysfs_zone_get_name(zone.val, subzone.val, subzone.set, name, sizeof(name)) > 0) {
        printf("%s\n", name);
      } else {
        ret = -errno;
        perror("Failed to get zone name");
      }
      break;
    case 'l':
      /* Get constraint power limit */
      if (!(ret = rapl_sysfs_constraint_get_power_limit_uw(zone.val, subzone.val, subzone.set, constraint.val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint power limit");
      }
      break;
    case 's':
      /* Get constraint time window */
      if (!(ret = rapl_sysfs_constraint_get_time_window_us(zone.val, subzone.val, subzone.set, constraint.val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint time window");
      }
      break;
    case 'U':
      /* Get constraint max power */
      if (!(ret = rapl_sysfs_constraint_get_max_power_uw(zone.val, subzone.val, subzone.set, constraint.val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint max power");
      }
      break;
    case 'u':
      /* Get constraint min power */
      if (!(ret = rapl_sysfs_constraint_get_min_power_uw(zone.val, subzone.val, subzone.set, constraint.val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint min power");
      }
      break;
    case 'T':
      /* Get constraint max time window */
      if (!(ret = rapl_sysfs_constraint_get_max_time_window_us(zone.val, subzone.val, subzone.set, constraint.val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint max time window");
      }
      break;
    case 't':
      /* Get constraint min time window */
      if (!(ret = rapl_sysfs_constraint_get_min_time_window_us(zone.val, subzone.val, subzone.set, constraint.val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint min time window");
      }
      break;
    case 'y':
      /* Get constraint name */
      if (rapl_sysfs_constraint_get_name(zone.val, subzone.val, subzone.set, constraint.val, name, sizeof(name)) > 0) {
        printf("%s\n", name);
      } else {
        ret = -errno;
        perror("Failed to get constraint name");
      }
      break;
    }
  } else if (zone.set || subzone.set || constraint.set) {
    /* Print summary of zone, subzone, or constraint */
    if (constraint.set) {
      /* print constraint */
      print_headers(zone.val, 1, subzone.val, subzone.set);
      analyze_constraint(zone.val, subzone.val, subzone.set, constraint.val, verbose);
    } else if (subzone.set) {
      /* print subzone */
      print_headers(zone.val, 1, 0, 0);
      analyze_zone(zone.val, subzone.val, subzone.set, verbose);
    } else {
      /* print zone */
      if (recurse) {
        analyze_zone_recurse(zone.val, verbose);
      } else {
        analyze_zone(zone.val, 0, 0, verbose);
      }
    }
  } else {
    /* print all zones */
    if ((ret = rapl_sysfs_zone_exists(0, 0, 0))) {
      fprintf(stderr, "No RAPL zones found\n");
    } else {
      analyze_all_zones_recurse(verbose);
    }
  }
  if (ret) {
    print_common_help();
  }

  return ret ? EXIT_FAILURE : EXIT_SUCCESS;
}
